File: Array/3Sum/solutionUsingTwoPointers.cpp
class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> results;
        int size = nums.size();

        for (int i = 0; i < size - 1; i++) {
            int left = i + 1;
            int right = size - 1;

            while (left < right) {
                if (nums[i] + nums[left] + nums[right] < 0) {
                    left++;
                } else if (nums[i] + nums[left] + nums[right] > 0) {
                    right--;
                } else {
                    results.push_back({nums[i], nums[left], nums[right]});
                    int leftIndexTemp = left;
                    int rightIndexTemp = right;

                    while (left < right && nums[left] == nums[leftIndexTemp]) {
                        left++;
                    }

                    while (left < right &&
                           nums[right] == nums[rightIndexTemp]) {
                        right--;
                    }
                }
            }

            while (i < size - 1 && nums[i] == nums[i + 1]) {
                i++;
            }
        }

        return results;
    }
};

File: Array/3Sum/solutionUsingTwoPointers.ts
function threeSum(nums: number[]): number[][] {
  nums.sort((a, b) => a - b);
  const result: number[][] = [];

  nums.forEach((value, i) => {
      if(i > 0 && nums[i] === nums[i - 1]) return;
      let left = i + 1;
      let right = nums.length - 1;
      
      while(left < right) {
          let sum: number = value + nums[left] + nums[right];
          if(sum < 0) left++;
          if(sum > 0) right--;
          if(sum === 0) {
              result.push([value, nums[left], nums[right]])
              left++;
              right--;
              while(left < right && nums[left] === nums[left - 1]) left++;
          }
      }

      return result;
  })

  return result;
};

// Input: nums = [-1,0,1,2,-1,-4] --> [-1,-1,0,1,2,4]
// Output: [[-1,-1,2],[-1,0,1]]